### PR TITLE
Add CSE; improve DCE and ConstProp

### PR DESCRIPTION
**Pull request by aswaterman**
Unit tests are forthcoming, pending a test template that's appropriate
for this kind of optimization pass.

### CODE REVIEW

**Comment by jackkoenig** [comment]
retest this please

**Comment by jackkoenig** [comment]
It seems that 6fe81557f893081a7365d66c1711c6f21ef16284 makes "Legalize" as it's currently written redundant. I guess we don't necessarily want Constant Prop to be required, but it does make Legalize pointless if it is run

File: src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    def recordNodes(s: Stmt): Stmt = s match {
+      case x: DefNode =>
+        nodes(x.name) = x.value
+        expressions.getOrElseUpdate(x.value, x.name)
+        x
+      case _ => s map recordNodes
+    }
+
+    def eliminateNodeRef(s: Expression): Expression = s match {
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala]: We generally use s for Stmt, e for Expression. Maybe it doesn't matter though

**Comment by aswaterman** [comment]
My inclination is also against requiring ConstProp for correctness.

File: src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala (hunk 1 above)
Comment (aswaterman) [inline src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala]: that's accidental, I'll fix it

**Comment by jackkoenig** [comment]
looks good to me, Travis might fail because of the concurrency issues but it passed before

**Comment by jackkoenig** [comment]
Travis is broken at the moment, this passed all FIRRTL and Chisel tests locally so I'm merging

**Pull request merged**